Alerting: Update rule API to address folders by UID

papagian opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

7. If your PR content should be added to the What's New document for the next major or minor release, add the **add to what's new** label to your PR. Note that you should add this label to the main PR that introduces the feature; do not add this label to smaller PRs for the feature.

-->

**What is this feature?**

Modifies the alerting rule API to expect a folder UID instead of folder title as namespace path parameter.

It's a draft: it requires additional frontend changes.

**Why do we need this feature?**

[Add a description of the problem the feature is trying to solve.]

**Who is this feature for?**

This is prerequisite for #63082

**Which issue(s) does this PR fix?**:

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #58184

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

# Release notice breaking change

The unstable alert rule API has been changed and now expects a folder UID instead of the folder title as namespace path parameter.
I addition to this, the responses that used to return the folder title now return `<folder parent UID>/<folder title>` to uniquely identify them.
Any consumers of the specific API should be appropriately adapted.

papagian committed:
Alerting: Update swagger definition

Add description for the namespace path parameters to highlight that
the folder UID is now expected.

papagian committed:
Change ruler API to expect the folder UID as namespace

papagian committed:
Update example requests

papagian committed:
fixup

papagian committed:
Fix tests

papagian committed:
Add TODO

VikaCep commented:
I started to work on these changes on the frontend but we noticed something odd. With these changes, we're showing rules twice, first under the folder UID and then under the folder name. This is because to check a rule actually exists, we must verify they're both present in the `grafana/api/v1/rules` response and also in `api/ruler/grafana/api/v1/rules`. 

![image](https://github.com/grafana/grafana/assets/6271380/83640d44-3047-4722-a953-33b17b8cd1ad)

The problem is `grafana/api/v1/rules` returns the folder name instead of the id in the `file` property 
![image](https://github.com/grafana/grafana/assets/6271380/d208292b-0d3d-4d34-9f02-4726fa849906)

while `api/ruler/grafana/api/v1/rules` returns the id as the key in the rule response
![image](https://github.com/grafana/grafana/assets/6271380/e17b438b-2f83-492f-8790-6ec291b84c79)

As the frontend tries to match the rules using these properties and they currently differ, they're being detected as different rules (and in a weird state, notice we show them either "creating" or "deleting"). In order to continue working both responses should have the same value.

**There is another important issue though**, if we changed the response in both requests to be only the folder UID, then the frontend wouldn't know how to obtain the folder name in order to display it in the UI and making yet another request to obtain this information would add complexity we'd rather want to avoid.
![image](https://github.com/grafana/grafana/assets/6271380/d8ab3d0e-856f-49af-9152-3bfa408912fa)

FYI @grobinson-grafana @gillesdemey

yuri-tceretian committed:
Merge branch 'up/main' into papagian/fix-alerting-58184

# Conflicts:
#	pkg/services/ngalert/api/api_ruler.go

yuri-tceretian commented:
I had to revert the changes done for GET apis because the struct `NamespaceConfigResponse` is used in Mimir\Loki paths and is supposed to have folder names as keys.

papagian committed:
Change lresponse

papagian committed:
Update swagger

papagian committed:
Fix tests

papagian commented:
@VikaCep I have pushed the changes (based on [this](https://docs.google.com/document/d/1tokkNXVWGOmTwdgUoGui3F_sWPdCMBVB2pOn8n_HLdk/edit#heading=h.htvcne9xxu3s))

<details><summary>Example `/api/ruler/grafana/api/v1/rules` response</summary>
<p>

```
{
	"baf2c548-5e1e-42e8-8fde-8320e50d801e": [
		{
			"name": "5m",
			"interval": "5m",
			"rules": [
				{
					"expr": "",
					"for": "5m",
					"grafana_alert": {
						"id": 1,
						"orgId": 1,
						"title": "alert rule under folder f0",
						"condition": "C",
						"data": [
							{
								"refId": "A",
								"queryType": "",
								"relativeTimeRange": {
									"from": 600,
									"to": 0
								},
								"datasourceUid": "PD8C576611E62080A",
								"model": {
									"intervalMs": 1000,
									"maxDataPoints": 43200,
									"refId": "A"
								}
							},
							{
								"refId": "B",
								"queryType": "",
								"relativeTimeRange": {
									"from": 0,
									"to": 0
								},
								"datasourceUid": "__expr__",
								"model": {
									"conditions": [
										{
											"evaluator": {
												"params": [],
												"type": "gt"
											},
											"operator": {
												"type": "and"
											},
											"query": {
												"params": [
													"B"
												]
											},
											"reducer": {
												"params": [],
												"type": "last"
											},
											"type": "query"
										}
									],
									"datasource": {
										"type": "__expr__",
										"uid": "__expr__"
									},
									"expression": "A",
									"intervalMs": 1000,
									"maxDataPoints": 43200,
									"reducer": "last",
									"refId": "B",
									"type": "reduce"
								}
							},
							{
								"refId": "C",
								"queryType": "",
								"relativeTimeRange": {
									"from": 0,
									"to": 0
								},
								"datasourceUid": "__expr__",
								"model": {
									"conditions": [
										{
											"evaluator": {
												"params": [
													0
												],
												"type": "gt"
											},
											"operator": {
												"type": "and"
											},
											"query": {
												"params": [
													"C"
												]
											},
											"reducer": {
												"params": [],
												"type": "last"
											},
											"type": "query"
										}
									],
									"datasource": {
										"type": "__expr__",
										"uid": "__expr__"
									},
									"expression": "B",
									"intervalMs": 1000,
									"maxDataPoints": 43200,
									"refId": "C",
									"type": "threshold"
								}
							}
						],
						"updated": "2023-09-08T11:36:07+03:00",
						"intervalSeconds": 300,
						"version": 1,
						"uid": "c2e8b23e-630a-485f-8b1e-0c092dd3471f",
						"namespace_uid": "baf2c548-5e1e-42e8-8fde-8320e50d801e",
						"namespace_id": 118,
						"namespace_title": "f0",
						"rule_group": "5m",
						"no_data_state": "NoData",
						"exec_err_state": "Error",
						"is_paused": false
					}
				}
			]
		}
	],
	"d9a88dcb-6737-454c-91a1-4213a8b00fdf": [
		{
			"name": "5m",
			"interval": "5m",
			"rules": [
				{
					"expr": "",
					"for": "5m",
					"grafana_alert": {
						"id": 2,
						"orgId": 1,
						"title": "alert rule under child",
						"condition": "C",
						"data": [
							{
								"refId": "A",
								"queryType": "",
								"relativeTimeRange": {
									"from": 600,
									"to": 0
								},
								"datasourceUid": "PD8C576611E62080A",
								"model": {
									"intervalMs": 1000,
									"maxDataPoints": 43200,
									"refId": "A"
								}
							},
							{
								"refId": "B",
								"queryType": "",
								"relativeTimeRange": {
									"from": 0,
									"to": 0
								},
								"datasourceUid": "__expr__",
								"model": {
									"conditions": [
										{
											"evaluator": {
												"params": [],
												"type": "gt"
											},
											"operator": {
												"type": "and"
											},
											"query": {
												"params": [
													"B"
												]
											},
											"reducer": {
												"params": [],
												"type": "last"
											},
											"type": "query"
										}
									],
									"datasource": {
										"type": "__expr__",
										"uid": "__expr__"
									},
									"expression": "A",
									"intervalMs": 1000,
									"maxDataPoints": 43200,
									"reducer": "last",
									"refId": "B",
									"type": "reduce"
								}
							},
							{
								"refId": "C",
								"queryType": "",
								"relativeTimeRange": {
									"from": 0,
									"to": 0
								},
								"datasourceUid": "__expr__",
								"model": {
									"conditions": [
										{
											"evaluator": {
												"params": [
													0
												],
												"type": "gt"
											},
											"operator": {
												"type": "and"
											},
											"query": {
												"params": [
													"C"
												]
											},
											"reducer": {
												"params": [],
												"type": "last"
											},
											"type": "query"
										}
									],
									"datasource": {
										"type": "__expr__",
										"uid": "__expr__"
									},
									"expression": "B",
									"intervalMs": 1000,
									"maxDataPoints": 43200,
									"refId": "C",
									"type": "threshold"
								}
							}
						],
						"updated": "2023-09-08T11:36:48+03:00",
						"intervalSeconds": 300,
						"version": 1,
						"uid": "f3819c8a-e770-42f4-b2c1-565d9631b117",
						"namespace_uid": "d9a88dcb-6737-454c-91a1-4213a8b00fdf",
						"namespace_id": 165,
						"namespace_title": "child",
						"rule_group": "5m",
						"no_data_state": "NoData",
						"exec_err_state": "Error",
						"is_paused": false
					}
				}
			]
		}
	]
}
```

</p>
</details>

papagian committed:
Revert "Fix tests"

This reverts commit c0e1255886fea328063e6e101a1bb2c4b0b7a7d2.

papagian committed:
Revert "Update swagger"

This reverts commit bdd1b40023fac68fcf59aa9c444f2b2a28d60048.

papagian committed:
Revert "Change lresponse"

This reverts commit ac878dd59f0d4bb95106a9976480d9043ee83142.

papagian committed:
Modify alert rule group response

papagian committed:
Regenerate swagger

papagian committed:
fixup

papagian commented:
@VikaCep I have changed it and implemented [the 2nd proposal](https://docs.google.com/document/d/1tokkNXVWGOmTwdgUoGui3F_sWPdCMBVB2pOn8n_HLdk/edit#heading=h.t7gn8ijy9fiw)

<details><summary>Example response</summary>
<p>

```
{
	"baf2c548-5e1e-42e8-8fde-8320e50d801e/f0": [
		{
			"name": "5m",
			"interval": "5m",
			"rules": [
				{
					"expr": "",
					"for": "5m",
					"grafana_alert": {
						"id": 1,
						"orgId": 1,
						"title": "alert rule under folder f0",
						"condition": "C",
						"data": [
							{
								"refId": "A",
								"queryType": "",
								"relativeTimeRange": {
									"from": 600,
									"to": 0
								},
								"datasourceUid": "PD8C576611E62080A",
								"model": {
									"intervalMs": 1000,
									"maxDataPoints": 43200,
									"refId": "A"
								}
							},
							{
								"refId": "B",
								"queryType": "",
								"relativeTimeRange": {
									"from": 0,
									"to": 0
								},
								"datasourceUid": "__expr__",
								"model": {
									"conditions": [
										{
											"evaluator": {
												"params": [],
												"type": "gt"
											},
											"operator": {
												"type": "and"
											},
											"query": {
												"params": [
													"B"
												]
											},
											"reducer": {
												"params": [],
												"type": "last"
											},
											"type": "query"
										}
									],
									"datasource": {
										"type": "__expr__",
										"uid": "__expr__"
									},
									"expression": "A",
									"intervalMs": 1000,
									"maxDataPoints": 43200,
									"reducer": "last",
									"refId": "B",
									"type": "reduce"
								}
							},
							{
								"refId": "C",
								"queryType": "",
								"relativeTimeRange": {
									"from": 0,
									"to": 0
								},
								"datasourceUid": "__expr__",
								"model": {
									"conditions": [
										{
											"evaluator": {
												"params": [
													0
												],
												"type": "gt"
											},
											"operator": {
												"type": "and"
											},
											"query": {
												"params": [
													"C"
												]
											},
											"reducer": {
												"params": [],
												"type": "last"
											},
											"type": "query"
										}
									],
									"datasource": {
										"type": "__expr__",
										"uid": "__expr__"
									},
									"expression": "B",
									"intervalMs": 1000,
									"maxDataPoints": 43200,
									"refId": "C",
									"type": "threshold"
								}
							}
						],
						"updated": "2023-09-08T11:36:07+03:00",
						"intervalSeconds": 300,
						"version": 1,
						"uid": "c2e8b23e-630a-485f-8b1e-0c092dd3471f",
						"namespace_uid": "baf2c548-5e1e-42e8-8fde-8320e50d801e",
						"namespace_id": 118,
						"rule_group": "5m",
						"no_data_state": "NoData",
						"exec_err_state": "Error",
						"is_paused": false
					}
				}
			]
		}
	],
	"d9a88dcb-6737-454c-91a1-4213a8b00fdf/child": [
		{
			"name": "5m",
			"interval": "5m",
			"rules": [
				{
					"expr": "",
					"for": "5m",
					"grafana_alert": {
						"id": 2,
						"orgId": 1,
						"title": "alert rule under child",
						"condition": "C",
						"data": [
							{
								"refId": "A",
								"queryType": "",
								"relativeTimeRange": {
									"from": 600,
									"to": 0
								},
								"datasourceUid": "PD8C576611E62080A",
								"model": {
									"intervalMs": 1000,
									"maxDataPoints": 43200,
									"refId": "A"
								}
							},
							{
								"refId": "B",
								"queryType": "",
								"relativeTimeRange": {
									"from": 0,
									"to": 0
								},
								"datasourceUid": "__expr__",
								"model": {
									"conditions": [
										{
											"evaluator": {
												"params": [],
												"type": "gt"
											},
											"operator": {
												"type": "and"
											},
											"query": {
												"params": [
													"B"
												]
											},
											"reducer": {
												"params": [],
												"type": "last"
											},
											"type": "query"
										}
									],
									"datasource": {
										"type": "__expr__",
										"uid": "__expr__"
									},
									"expression": "A",
									"intervalMs": 1000,
									"maxDataPoints": 43200,
									"reducer": "last",
									"refId": "B",
									"type": "reduce"
								}
							},
							{
								"refId": "C",
								"queryType": "",
								"relativeTimeRange": {
									"from": 0,
									"to": 0
								},
								"datasourceUid": "__expr__",
								"model": {
									"conditions": [
										{
											"evaluator": {
												"params": [
													0
												],
												"type": "gt"
											},
											"operator": {
												"type": "and"
											},
											"query": {
												"params": [
													"C"
												]
											},
											"reducer": {
												"params": [],
												"type": "last"
											},
											"type": "query"
										}
									],
									"datasource": {
										"type": "__expr__",
										"uid": "__expr__"
									},
									"expression": "B",
									"intervalMs": 1000,
									"maxDataPoints": 43200,
									"refId": "C",
									"type": "threshold"
								}
							}
						],
						"updated": "2023-09-08T11:36:48+03:00",
						"intervalSeconds": 300,
						"version": 1,
						"uid": "f3819c8a-e770-42f4-b2c1-565d9631b117",
						"namespace_uid": "d9a88dcb-6737-454c-91a1-4213a8b00fdf",
						"namespace_id": 165,
						"rule_group": "5m",
						"no_data_state": "NoData",
						"exec_err_state": "Error",
						"is_paused": false
					}
				}
			]
		}
	]
}
```

</p>
</details>

papagian committed:
Modify FIle field in /api/prometheus/grafana/api/v1/rules

papagian committed:
Merge remote-tracking branch 'origin/main' into papagian/fix-alerting-58184

papagian committed:
fix ruler export

papagian committed:
Modify folder in responses to be formatted as <parent UID>/<title>

yuri-tceretian commented on pkg/services/ngalert/api/api_ruler_export.go:
```diff
@@ -13,8 +13,8 @@ import (
 // Can return 403 StatusForbidden if user is not authorized to read folder `namespaceTitle`
-func (srv RulerSrv) ExportFromPayload(c *contextmodel.ReqContext, ruleGroupConfig apimodels.PostableRuleGroupConfig, namespaceTitle string) response.Response {
-	namespace, err := srv.store.GetNamespaceByTitle(c.Req.Context(), namespaceTitle, c.SignedInUser.OrgID, c.SignedInUser)
+func (srv RulerSrv) ExportFromPayload(c *contextmodel.ReqContext, ruleGroupConfig apimodels.PostableRuleGroupConfig, namespaceUID string) response.Response {
```
please change the comment too.

yuri-tceretian commented on pkg/services/ngalert/api/persist.go:
```diff
@@ -10,8 +10,9 @@ import (
 type RuleStore interface {
+	// TODO after deprecating namespace_id field in GettableGrafanaRule we can simplify this interface
+	// by returning map[string]struct{} instead of map[string]*folder.Folder
 	GetUserVisibleNamespaces(context.Context, int64, *user.SignedInUser) (map[string]*folder.Folder, error)
```
but we still need folder title, right?

yuri-tceretian commented on pkg/services/ngalert/api/tooling/definitions/cortex-ruler.go:
```diff
@@ -152,6 +152,7 @@ import (
 
 // swagger:parameters RoutePostNameRulesConfig RoutePostNameGrafanaRulesConfig RoutePostRulesGroupForExport
 type NamespaceConfig struct {
+	// The UID of the rule folder
```
IIRC this is shared between Lotex and Grafana APIs, and in the former case it's name, not UID.

yuri-tceretian commented on pkg/services/ngalert/api/tooling/definitions/cortex-ruler.go:
```diff
@@ -402,22 +405,23 @@ type PostableGrafanaRule struct {
+	UID             string       `json:"uid" yaml:"uid"`
+	NamespaceUID    string       `json:"namespace_uid" yaml:"namespace_uid"`
+	// Deprecated: it will be removed in a future release in favor of NamespaceUID
+	NamespaceID  int64               `json:"namespace_id" yaml:"namespace_id"`
```
@grafana/alerting-frontend I wonder if it's possible to get rid of it now.

VikaCep replied:
We're not using it, we should be ok to remove it.

alexweav replied:
This has been removed in a separate PR. I've rebased this PR on top, we no longer need to worry about this here.

yuri-tceretian commented on pkg/tests/api/alerting/api_ruler_test.go:
```diff
@@ -82,7 +82,7 @@ func TestIntegrationAlertRulePermissions(t *testing.T) {
 		expectedGetNamespaceResponseBody := `
 		{
-		   "folder1":[
+		   "/folder1":[
```
It would be great to update the tests to run with one nested folder.

zserge commented on pkg/services/ngalert/api/api_ruler.go:
```diff
@@ -27,6 +28,8 @@ import (
 	"github.com/grafana/grafana/pkg/util"
 )
 
+const NAMESPACE_SEPARATOR = "/"
```
If we only use forward slashes - perhaps we could use path.Join vs strings.Join?

papagian replied:
on second thought I think that this is better (no change for existing flat folders) so I have followed your suggestion

papagian committed:
Merge remote-tracking branch 'origin/main' into papagian/fix-alerting-58184

papagian committed:
Fix test

papagian commented:
blocked by #76504: alerting using [search](https://github.com/grafana/grafana/blob/d72ec22ec290d7a729f601dcfdf50fcd03994dca/pkg/services/ngalert/store/alert_rule.go#L448) for fetching the folder details

papagian committed:
Merge remote-tracking branch 'origin/main' into papagian/fix-alerting-58184

papagian committed:
WIP add alerting test with nested folders

papagian committed:
Merge remote-tracking branch 'origin/main' into papagian/fix-alerting-58184

papagian committed:
Fix test for nested folders

papagian committed:
Ignore lint failure due to deprecated fields

VikaCep commented:
From https://docs.google.com/document/d/1tokkNXVWGOmTwdgUoGui3F_sWPdCMBVB2pOn8n_HLdk/edit#heading=h.t7gn8ijy9fiw (proposal 2) it seems now the frontend won't need to change the way we construct the namespace parameter for the rule API, am I understanding this correctly?

If so then I am confused as I'm getting 404 errors from the API when trying to perform operations such as saving a rule / reordering them:

Saving a new rule:
![2023-10-25 16 46 09](https://github.com/grafana/grafana/assets/6271380/75207656-5cc9-4a30-b372-f4478b812753)

Reordering:
![2023-10-25 16 49 00](https://github.com/grafana/grafana/assets/6271380/5d4cd799-68f8-47b4-909b-ef818284fabf)

yuri-tceretian commented:
@VikaCep the UI should refer to folder by UID in all POST/PUT requests. The document you refer to discusses only how GET APIs work.

yuri-tceretian commented on pkg/tests/api/alerting/api_ruler_test.go:
```diff
@@ -240,15 +309,18 @@ func TestIntegrationAlertRulePermissions(t *testing.T) {
+		// remove permissions for folder2
 		removeFolderPermission(t, permissionsStore, 1, userID, org.RoleEditor, "folder2")
+		// remove permissions for subfolder (inherits from folder1)
+		removeFolderPermission(t, permissionsStore, 1, userID, org.RoleEditor, "subfolder")
```
I think this is ok but I would really love to see it as a separate sub-test.

papagian replied:
done in 060b541f67b86ff77480680d05c05dd03045d7f2

yuri-tceretian reviewed (commented):
Back-end changes looks good to me but in my opinion, it will be much safer if we have changes in the back-end and front-end in the same PR.
@VikaCep do you think you can commit FE changes to this PR?

Also, and I think this is not problem of this PR but nested folders feature in general: the file provisioning of alert rules still refers to folders by name. It is not clear how it is supposed to work with nested folders. @papagian, do you happen to know if dashboards provisioning have figured out this problem?

papagian commented on pkg/tests/api/alerting/api_ruler_test.go:
```diff
@@ -188,6 +227,7 @@ func TestIntegrationAlertRulePermissions(t *testing.T) {
 
 			require.Equal(t, "folder1", allExport.Groups[0].Folder)
 			require.Equal(t, "folder2", allExport.Groups[1].Folder)
+			require.Equal(t, "subfolder", allExport.Groups[2].Folder)
```
I have opened another PR for that: [Alerting: Export and provisioning rules into subfolders](https://github.com/grafana/grafana/pull/77450)

papagian committed:
Merge remote-tracking branch 'origin/main' into papagian/fix-alerting-58184

papagian committed:
fix tests

papagian committed:
Update OpenAPI specs

papagian committed:
Merge remote-tracking branch 'origin/main' into papagian/fix-alerting-58184

papagian commented on pkg/tests/api/alerting/test-data/rulegroup-3-export.json:
```diff
@@ -0,0 +1,75 @@
+    {
+      "orgId": 1,
+      "name": "Group3",
+      "folder": "<dynamic>",
```
same as below I have opened another PR for that: [Alerting: Export and provisioning rules into subfolders](https://github.com/grafana/grafana/pull/77450)

papagian committed:
Update comment

papagian committed:
Split tests and do not add remove leading slash in fullpath

papagian committed:
fixup

papagian committed:
make lint-go

yuri-tceretian commented on pkg/services/ngalert/api/api_ruler.go:
```diff
@@ -578,5 +578,8 @@ func (srv RulerSrv) searchAuthorizedAlertRules(ctx context.Context, c *contextmo
 func getNamespaceKey(namespace *folder.Folder) string {
+	if namespace.ParentUID == "" {
+		return namespace.Title
+	}
```
I think we should still return the leading slash and UI should strip it. This will simplify logic on the front end. Without the leading slash a folder title like "My Folder / Alerts" would could be treated by UI as a nested folder. 
However, if we return a folder like "/My Folder / Alerts" the UI can always treat the first slash as a folder separator. 

Alternative, is to escape slashes in the `namespace.Title`

papagian replied:
I think I prefer the alternative to escape slashes in the `namespace.Title` because the other (including the leading slash) would introduce bigger breaking changes when [provisioning to folders](https://github.com/grafana/grafana/pull/77450) (it would require users to modify all there existing provisioning configurations and add the leading slash)

yuri-tceretian commented:
Great job, Sofia! Thanks for addressing my comments. From the backend perspective, the changes are LGTM after my comment about leading slash is addressed.

papagian committed:
Apply suggestion from code review

papagian committed:
Merge remote-tracking branch 'origin/main' into papagian/fix-alerting-58184

VikaCep committed:
Alerting: use folder UID instead of title in rule API (#77166)

Co-authored-by: Sonia Aguilar <soniaaguilarpeiron@gmail.com>

yuri-tceretian commented:
/deploy-to-hg

papagian committed:
Merge remote-tracking branch 'origin/main' into papagian/fix-alerting-58184

yuri-tceretian commented:
/deploy-to-hg

yuri-tceretian commented:
I poked UI around with and without nestedFolder flag enabled and everything works fine for me.

LGTM! Thank you so much @papagian and @VikaCep for doing such big amount of work!

VikaCep reviewed (approved):
🚀 🚀 🚀

papagian commented:
> Apparently, we need to update some pieces that are overridden by cloud alerting scheduler (see remoteruler code in GE). It overrides path `/api/prometheus/grafana/api/v1/rules` and therefore the emphemeral instance that is running now is broken: one api call returns nested folder correctly but the overridden - returns regular title. It happens only if nested folders are enabled.

@yuri-tceretian thanks for testing it thoroughly! I didn't know anything about the remote ruler override: I assume we have to wait for the GE fix before merging it, right?

yuri-tceretian commented:
> I assume we have to wait for the GE fix before merging it, right?

Yes, otherwise we get broken cloud build.

yuri-tceretian commented:
After some digging, I think we only need to change this piece of code
 https://github.com/grafana/grafana/blob/6d4625ad52b2e2ad3b2b88e394f69e8fde83c04b/pkg/services/ngalert/store/alert_rule.go#L566-L580

we use it to serve that API but it is also used in the scheduler, which adds folder title to alert as a label `grafana_folder`.

papagian commented:
> After some digging, I think we only need to change this piece of code to return folders in a correct format
> 
> https://github.com/grafana/grafana/blob/6d4625ad52b2e2ad3b2b88e394f69e8fde83c04b/pkg/services/ngalert/store/alert_rule.go#L566-L580
> 
> we use it to serve that API but it is also used in the scheduler, which adds folder title to alert as a label `grafana_folder`.

do you mean that we only need to format the values of `ResultFoldersTitles` map like this?
https://github.com/grafana/grafana/blob/9fc1d506a971f77bf33a5a7983c39f95fba1c70b/pkg/services/ngalert/api/api_ruler.go#L583

alexweav committed:
Merge branch 'main' into papagian/fix-alerting-58184

alexweav committed:
Merge branch 'main' into papagian/fix-alerting-58184

alexweav committed:
Drop a few more latent uses of namespace_id

yuri-tceretian committed:
move getNamespaceKey to models package

yuri-tceretian committed:
switch GetAlertRulesForScheduling to use folder table

yuri-tceretian committed:
update GetAlertRulesForScheduling to return folder titles in format `parent_uid/title`.

yuri-tceretian committed:
Merge branch 'up/main' into papagian/fix-alerting-58184

# Conflicts:
#	public/app/features/alerting/unified/components/rule-viewer/RuleViewer.v1.tsx

yuri-tceretian committed:
fi tests

yuri-tceretian committed:
add tests for GetAlertRulesForScheduling when parent uid

yuri-tceretian committed:
fix integration tests after merge

yuri-tceretian committed:
Merge branch 'up/main' into papagian/fix-alerting-58184

# Conflicts:
#	pkg/services/ngalert/store/alert_rule.go

yuri-tceretian committed:
fix test after merge

yuri-tceretian commented:
/deploy-to-hg

yuri-tceretian commented:
/deploy-to-hg

yuri-tceretian committed:
Merge branch 'up/main' into papagian/fix-alerting-58184

yuri-tceretian commented:
/deploy-to-hg

yuri-tceretian commented:
/deploy-to-hg --enterprise-ref papagian/fix-alerting-58184

yuri-tceretian commented:
/deploy-to-hg --enterprise-ref papagian/fix-alerting-58184

yuri-tceretian reviewed (approved):
LGTM. We need to merge PR in GE and this one together.

yuri-tceretian committed:
change format of the namespace to JSON array

this is needed for forward compatibility, when we migrate to full paths

yuri-tceretian committed:
Merge branch 'up/main' into papagian/fix-alerting-58184

gillesdemey committed:
update EF code to decode nested folder

gillesdemey reviewed (approved):
Latest FE changes LGTM!

alexweav commented:
LGTM
